### PR TITLE
fix(smart-contracts): `receive()` function safety upgrade

### PR DESCRIPTION
**Pull request by abdulhakim-altunkaya**
# Adding a require() function inside receive() function to prevent zero or negative transactions. A safety precaution against malicious or unexpected behaviour.

# receive() function can be exloited or it can cause unexpected behaviour.

Fixes #
Refs #

# Checklist:

- [ ] 1 PR, 1 purpose: my Pull Request applies to a single purpose
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have updated the docs to reflect my changes if applicable
- [ ] I have added tests (and stories for frontend components) that prove my fix is effective or that my feature works
- [ ] I have performed a self-review of my own code
- [ ] If my code involves visual changes, I am adding applicable screenshots to this thread

### CODE REVIEW

**Comment by abdulhakim-altunkaya** [comment]
CLA signed

**Comment by abdulhakim-altunkaya** [comment]
Hello Clement,

Sure, I agree with you about customs errors. They are better alternatives
compared to require. I used "require" just to draw attention to that small
improvement area.

And about the exploit, receive() function can be triggered even with zero
value transactions. So, when a zero-value transaction is sent to a
contract, it still requires a certain amount of gas to be processed by the
network. This means that if an attacker sends a large number of zero-value
transactions to a contract, it can quickly consume all available gas
for the block, preventing other transactions from being executed. This
attack is called as Denial-of-Service (DoS) attack.

In other words, if an attacker sends a large number of zero value
transactions to a contract, the gas consumed by these transactions will add
up and eventually exceed the maximum gas limit for the current block. This
will cause the remaining transactions in the block to fail and any changes
made by those transactions will be undone. In this way, the attacker can
cause a Denial-of-Service (DoS) attack by preventing other transactions
from being executed.

Also, I could not see a function (at least in Unlock.sol) which will later
help you to withdraw the funds coming with "receive()"

**Comment by clemsos** [comment]
Yes, the network DDOS makes sense. Actually the `receive` function is not in use anymore (as you noticed) - as the logic to unwrap wETH has been moved to another standalone contract. Thanks for the scrutiny, that's very useful !

Closing this in favour of #11511 11511

**Review by clemsos — APPROVED** [review]
LG!
